static/app/views/preprod/buildDetails/header/useBuildDetailsActions.tsx

```diff
@@ -52,19 +52,66 @@ export function useBuildDetailsActions({
     });
   };
 
-  const handleDownloadAction = () => {
+  const handleDownloadAction = async () => {
+    const downloadUrl = `/api/0/internal/${organization.slug}/${projectId}/files/preprodartifacts/${artifactId}/`;
+
     try {
-      const downloadUrl = `/api/0/internal/${organization.slug}/${projectId}/files/preprodartifacts/${artifactId}/`;
+      const response = await fetch(downloadUrl, {
+        method: 'HEAD',
+        credentials: 'include',
+      });
+
+      if (!response.ok) {
+        let errorMessage = `Download failed (${response.status})`;
+
+        try {
+          const errorResponse = await fetch(downloadUrl, {
+            method: 'GET',
+            credentials: 'include',
+          });
+
+          if (!errorResponse.ok) {
+            const errorText = await errorResponse.text();
+            try {
+              const errorJson = JSON.parse(errorText);
+              if (errorJson.detail) {
+                if (typeof errorJson.detail === 'string') {
+                  errorMessage = errorJson.detail;
+                } else if (errorJson.detail.message) {
+                  errorMessage = errorJson.detail.message;
+                } else if (errorJson.detail.code) {
+                  errorMessage = `${errorJson.detail.code}: ${errorJson.detail.message || 'Authentication required'}`;
+                }
+              }
+            } catch {
+              errorMessage = errorText || errorMessage;
+            }
+          }
+        } catch {
```

Comment:
I think this catch could just be around `const errorResponse = await fetch(downloadUrl, ...);`
That way every thing is a bit less nested and easier to follow.